exonum-java-binding-core/rust/src/utils/jni_cache.rs

```diff
@@ -0,0 +1,158 @@
+use jni::{
+    objects::{GlobalRef, JMethodID},
+    sys::{jint, JNI_VERSION_1_8},
+    JNIEnv, JavaVM,
+};
+use std::os::raw::c_void;
+
+static mut OBJECT_CACHED_GET_CLASS: Option<JMethodID> = None;
+static mut CLASS_CACHED_GET_NAME: Option<JMethodID> = None;
+static mut THROWABLE_CACHED_GET_MESSAGE: Option<JMethodID> = None;
+
+static mut UTA_CACHED_EXECUTE: Option<JMethodID> = None;
+static mut UTA_CACHED_INFO: Option<JMethodID> = None;
+static mut UTA_CACHED_VERIFY: Option<JMethodID> = None;
+
+static mut USA_CACHED_STATE_HASHES: Option<JMethodID> = None;
+static mut USA_CACHED_CONVERT_TRANSACTION: Option<JMethodID> = None;
+
+static mut CLASS_CACHED_ERROR: Option<GlobalRef> = None;
+static mut CLASS_CACHED_TEE: Option<GlobalRef> = None;
+
+#[allow(non_snake_case)]
+#[no_mangle]
+pub extern "system" fn JNI_OnLoad(vm: JavaVM, _: *mut c_void) -> jint {
+    let env = vm.get_env().expect("Cannot get reference to the JNIEnv");
+    cache_methods(&env);
+
+    JNI_VERSION_1_8
+}
+
+/// Caches all required classes and methods ids.
+pub fn cache_methods(env: &JNIEnv) {
+    unsafe {
+        OBJECT_CACHED_GET_CLASS =
+            get_method_id(&env, "java/lang/Object", "getClass", "()Ljava/lang/Class;");
+        CLASS_CACHED_GET_NAME =
+            get_method_id(&env, "java/lang/Class", "getName", "()Ljava/lang/String;");
+        THROWABLE_CACHED_GET_MESSAGE = get_method_id(
+            &env,
+            "java/lang/Throwable",
+            "getMessage",
+            "()Ljava/lang/String;",
+        );
+        UTA_CACHED_EXECUTE = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserTransactionAdapter",
+            "execute",
+            "(J)V",
+        );
+        UTA_CACHED_INFO = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserTransactionAdapter",
+            "info",
+            "()Ljava/lang/String;",
+        );
+        UTA_CACHED_VERIFY = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserTransactionAdapter",
+            "isValid",
+            "()Z",
+        );
+        USA_CACHED_STATE_HASHES = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserServiceAdapter",
+            "getStateHashes",
+            "(J)[[B",
+        );
+        USA_CACHED_CONVERT_TRANSACTION = get_method_id(
+            &env,
+            "com/exonum/binding/service/adapters/UserServiceAdapter",
+            "convertTransaction",
+            "([B)Lcom/exonum/binding/service/adapters/UserTransactionAdapter;",
+        );
+        CLASS_CACHED_ERROR = env
+            .new_global_ref(env.find_class("java/lang/Error").unwrap().into())
+            .ok();
+        CLASS_CACHED_TEE = env
+            .new_global_ref(
+                env.find_class("com/exonum/binding/transaction/TransactionExecutionException")
+                    .unwrap()
+                    .into(),
```

Comment:
Formatting here is like a complete disaster 😕

Reply:
But it comes from the regular format command :)